### PR TITLE
Issue #1405 Merge disjoint c8y crates into unified c8y_api crate

### PR DESCRIPTION
**Pull request by albinsuresh**
## Proposed changes

Merge `c8y_smartrest` crate into the unified `c8y_api` crate. The `c8y_translator` crate will be merged into the `c8y_mapper` in a follow-up PR as it is independent from this work.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by didier-wenzek — APPROVED** [review]
Approved as first step toward https://github.com/thin-edge/thin-edge.io/issues/1405.

File: crates/core/c8y_api/src/smartrest/topic.rs [hunk 1]
```diff
@@ -3,24 +3,44 @@ use agent_interface::TopicError;
...
+    //     match self {
+    //         Self::SmartRestRequest => SMARTREST_SUBSCRIBE_TOPIC,
+    //         Self::SmartRestResponse(None) => SMARTREST_PUBLISH_TOPIC,
+    //         Self::SmartRestResponse(Some(child_id)) => {
+    //             let child_topic = format!("{}/{}", SMARTREST_PUBLISH_TOPIC, child_id);
+    //             child_topic.as_str()
+    //         }
+    //         Self::OperationTopic(name) => name.as_str(),
+    //     }
+    // }
```
Comment (didier-wenzek) [inline crates/core/c8y_api/src/smartrest/topic.rs:27]: To be removed if unused,

File: crates/core/tedge_mapper/src/c8y/converter.rs [hunk 2]
```diff
@@ -472,7 +471,7 @@ async fn parse_c8y_topics(
                 ..
             },
         ) => {
-            let topic = C8yTopic::SmartRestResponse.to_topic()?;
+            let topic = C8yTopic::SmartRestResponse(None).to_topic()?;
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/converter.rs:474]: When I see such a `SmartRestResponse(None)` I wonder what could have been this `None`. It's not obvious that this a child device id.

Why not something similar to:

```
pub enum C8yTopic {
    SmartRestRequest,
    SmartRestResponse,
    ChildSmartRestResponse(String),
    OperationTopic(String),
}
```

This also raises another question (that is beyond that PR):
* Why a response topic for the child devices, but nothing similar for the requests and the operations?

**Review by didier-wenzek — COMMENTED** [review]
Unfortunately, you have new conflicts :-(

File: crates/core/tedge_mapper/src/c8y/converter.rs (hunk 2 above)
Comment (albinsuresh) [reply crates/core/tedge_mapper/src/c8y/converter.rs:474]: I agree that having a new variant is better than overloading the existing one. Regarding the lack of "requests" and "operations" for child devices, it is as designed by Cumulocity. There's no `s/ds/{child-id}` topic as the requests for the parent device as well as child devices are received on `s/ds` topic with the `child-id` in the payload.

Although inconsistent with `s/us/{child-id}`, this design is beneficial to us as we don't have to keep restarting the msoquitto bridge to subscribe to new child topics whenever a new child gets added.

**Pull request merged**